File: _book/source/_posts/Algorithm/1.leetcode/tmp/39.组合总和.py
#
# @lc app=leetcode.cn id=39 lang=python3
#
# [39] 组合总和
#

from typing import List


# @lc code=start
class Solution:
    def combinationSum(self, candidates: List[int],
                       target: int) -> List[List[int]]:
        size = len(candidates)
        if size < 0:
            return []
        print(candidates.sort())
        path = []
        res = []
        self._find_path(target, path, res, candidates, 0, size)
        return res

    def _find_path(self, target, path, res, candidates, begin, size):
        if target == 0:
            res.append(path.copy())
        else:
            for i in range(begin, size):
                left_num = target - candidates[i]
                if left_num < 0:
                    # 剩余值为负数， 剪枝
                    break
                # 否则把当前值加入路径
                path.append(candidates[i])
                self._find_path(left_num, path, res, candidates, i, size)
                path.pop()


# @lc code=end
tst = Solution()
print(tst.combinationSum([2, 3, 6, 7], 7))

File: source/_posts/Algorithm/算法专题/回溯/39.组合总和.java
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/*
 * @lc app=leetcode.cn id=39 lang=java
 * 
 * [39] 组合总和
 */

// @lc code=start
class Solution {
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> res = new ArrayList<>();
        Arrays.sort(candidates);
        backtrack(res, new ArrayList<>(), candidates, target, 0);
        return res;
    }

    void backtrack(List<List<Integer>> res, List<Integer> tempList, int[] candidates, int remain, int start) {
        if (remain < 0) {
            return;
        } else if (remain == 0) {
            res.add(new ArrayList<>(tempList));
        } else {
            for (int i = start; i < candidates.length; i++) {
                tempList.add(candidates[i]);
                backtrack(res, tempList, candidates, remain - candidates[i], i);
                tempList.remove(tempList.size() - 1);
            }
        }
    }

    public static void main(String[] args) {
        Solution solution = new Solution();
        int[] candidates = new int[] { 2, 3, 6, 7 };
        solution.combinationSum(candidates, 7);
    }
}
// @lc code=end
